shipyard-controller/nats/receiver.go

```diff
@@ -0,0 +1,145 @@
+	ctx            context.Context
+	jetStream      nats.JetStreamContext
+}
+
+func NewNatsConnectionHandler(ctx context.Context, natsURL string, messageHandler IKeptnNatsMessageHandler) *NatsConnectionHandler {
+	return &NatsConnectionHandler{natsURL: natsURL, messageHandler: messageHandler, ctx: ctx}
+}
+
+func (nch *NatsConnectionHandler) RemoveAllSubscriptions() {
+	for _, sub := range nch.subscriptions {
+		_ = sub.Unsubscribe()
+		logger.Infof("Unsubscribed from NATS topic: %s", sub.subscription.Subject)
+	}
+	nch.subscriptions = nch.subscriptions[:0]
+}
+
+// SubscribeToTopics expresses interest in the given subject(s) on the NATS message broker.
+func (nch *NatsConnectionHandler) SubscribeToTopics(topics []string) error {
+	if nch.natsURL == "" {
+		return errors.New("no PubSub URL defined")
+	}
+
+	if nch.natsConnection == nil || !nch.natsConnection.IsConnected() {
+		var err error
+		nch.RemoveAllSubscriptions()
+
+		nch.natsConnection.Close()
+		logger.Infof("Connecting to NATS server at %s ...", nch.natsURL)
+		nch.natsConnection, err = nats.Connect(nch.natsURL)
+
+		if err != nil {
+			return errors.New("failed to create NATS connection: " + err.Error())
+		}
+
+		err = nch.setupJetStreamContext(topics)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(topics) > 0 && !IsEqual(nch.topics, topics) {
+		nch.RemoveAllSubscriptions()
+		nch.topics = topics
+
+		for _, topic := range nch.topics {
+			subscription := NewPullSubscription(nch.ctx, queueGroup, topic, nch.jetStream, nch.messageHandler.Process)
+			if err := subscription.Activate(); err != nil {
+				return fmt.Errorf("could not start subscription: %s", err.Error())
+			}
+			nch.subscriptions = append(nch.subscriptions, subscription)
+		}
+	}
+	return nil
+}
+
+func (nch *NatsConnectionHandler) setupJetStreamContext(topics []string) error {
+	js, err := nch.natsConnection.JetStream()
+	if err != nil {
+		return fmt.Errorf("failed to create nats jetStream context: %s", err.Error())
+	}
+
+	stream, err := js.StreamInfo(streamName)
+	if err != nil && err != nats.ErrStreamNotFound {
+		return fmt.Errorf("failed to retrieve stream info: %s", err.Error())
+	}
+	if stream == nil {
+		logger.Infof("creating stream %q", streamName)
+		_, err = js.AddStream(getShipyardStreamConfig(topics))
+		if err != nil {
+			return fmt.Errorf("failed to add stream: %s", err.Error())
+		}
+	} else {
+		_, err = js.UpdateStream(getShipyardStreamConfig(topics))
+		if err != nil {
+			return fmt.Errorf("failed to update stream: %s", err.Error())
+		}
+	}
+	nch.jetStream = js
+	return nil
+}
+
+func getShipyardStreamConfig(topics []string) *nats.StreamConfig {
+	return &nats.StreamConfig{
+		Name:     streamName,
+		Subjects: topics,
+	}
+}
+
+func IsEqual(a1 []string, a2 []string) bool {
+	sort.Strings(a1)
+	sort.Strings(a2)
+	if len(a1) == len(a2) {
+		for i, v := range a1 {
+			if v != a2[i] {
+				return false
+			}
+		}
+	} else {
+		return false
+	}
+	return true
+}
```

Comment:
Why not use something like https://pkg.go.dev/reflect#DeepEqual ?

Reply:
true, will use that